src/rust/engine/src/intrinsics.rs

```diff
@@ -280,33 +280,39 @@ fn merge_digests_request_to_digest(
   .boxed()
 }
 
-fn url_to_fetch_to_snapshot(
+fn url_to_fetch_to_digest(
   context: Context,
   mut args: Vec<Value>,
 ) -> BoxFuture<'static, NodeResult<Value>> {
   let core = context.core.clone();
   async move {
+    let key = externs::acquire_key_for(args.pop().unwrap()).map_err(|e| format!("{:?}", e))?;
     let snapshot = context
-      .get(DownloadedFile(externs::acquire_key_for(
-        args.pop().unwrap(),
-      )?))
-      .await?;
-    Ok(Snapshot::store_snapshot(&core, &snapshot).map_err(|err| throw(&err))?)
+      .get(DownloadedFile(key))
+      .await
+      .map_err(|e| format!("{:?}", e))?;
+    let res: Result<_, String> = Ok(Snapshot::store_directory(&core, &snapshot.digest));
+    res
   }
+  .map_err(|err: String| throw(&err))
   .boxed()
 }
 
-fn path_globs_to_snapshot(
+fn path_globs_to_digest(
   context: Context,
   mut args: Vec<Value>,
 ) -> BoxFuture<'static, NodeResult<Value>> {
   let core = context.core.clone();
   async move {
+    let key = externs::acquire_key_for(args.pop().unwrap()).map_err(|e| format!("{:?}", e))?;
     let snapshot = context
-      .get(Snapshot(externs::acquire_key_for(args.pop().unwrap())?))
-      .await?;
-    Ok(Snapshot::store_snapshot(&core, &snapshot).map_err(|err| throw(&err))?)
+      .get(Snapshot(key))
+      .await
+      .map_err(|e| format!("{:?}", e))?;
+    let res: Result<_, String> = Ok(Snapshot::store_directory(&core, &snapshot.digest));
+    res
```

Comment:
Currently I think that we will still actually cache the Snapshot in memory here. To fix that, you'd need to change the implementation of the `Snapshot` Node: https://github.com/pantsbuild/pants/blob/d63b29cfdd6d666fb6475cd2362c21c96a931ac8/src/rust/engine/src/nodes.rs#L597-L607

... to have `type Item = Digest;`, probably. That will avoid keeping in memory a bunch of stuff that there is a separate explicit operation to load.